sdks/python/apache_beam/runners/direct/sdf_direct_runner.py

```diff
@@ -0,0 +1,264 @@
+  """A transform override for ProcessElements transform."""
+
+  def get_matcher(self):
+    def _matcher(applied_ptransform):
+      return isinstance(
+          applied_ptransform.transform, ProcessKeyedElements)
+
+    return _matcher
+
+  def get_replacement_transform(self, ptransform):
+    return ProcessKeyedElementsViaKeyedWorkItems(ptransform)
+
+
+class ProcessKeyedElementsViaKeyedWorkItems(PTransform):
+  """A transform that processes Splittable DoFn input via KeyedWorkItems.
+  """
+
+  def __init__(self, process_keyed_elements_transform):
+    self._process_keyed_elements_transform = process_keyed_elements_transform
+
+  def expand(self, pcoll):
+    return pcoll | beam.core.GroupByKey() | ProcessElements(
+        self._process_keyed_elements_transform)
+
+
+class ProcessElements(PTransform):
+  """Processes keyed input via Splittable DoFn objects."""
+
+  def __init__(self, process_keyed_elements_transform):
+    self._process_keyed_elements_transform = process_keyed_elements_transform
+    self.sdf = self._process_keyed_elements_transform.sdf
+
+  def expand(self, pcoll):
+    return pvalue.PCollection(pcoll.pipeline)
+
+  def new_process_fn(self, sdf):
+    return ProcessFn(
+        sdf,
+        self._process_keyed_elements_transform.ptransform_args,
+        self._process_keyed_elements_transform.ptransform_kwargs)
+
+
+class ProcessFn(beam.DoFn):
+
+  def __init__(
+      self, sdf, args_for_invoker, kwargs_for_invoker):
+    self.sdf = sdf
+    self._element_tag = _ValueStateTag('element')
+    self._restriction_tag = _ValueStateTag('restriction')
+    self.watermark_hold_tag = _ValueStateTag('watermark_hold')
+    self._process_element_invoker = None
+
+    self.sdf_invoker = DoFnInvoker.create_invoker(
+        DoFnSignature(self.sdf), context=DoFnContext('unused_context'),
+        input_args=args_for_invoker, input_kwargs=kwargs_for_invoker)
+
+  def set_step_context(self, step_context):
+    assert isinstance(step_context, DirectStepContext)
+    self._step_context = step_context
+
+  def set_process_element_invoker(self, process_element_invoker):
+    assert isinstance(process_element_invoker, SDFProcessElementInvoker)
+    self._process_element_invoker = process_element_invoker
+
+  def start_bundle(self):
+    # TODO: support start_bundle() method for SDFs.
+    pass
+
+  def finish_bundle(self):
+    # TODO: support finish_bundle() method for SDFs.
+    pass
+
+  def process(self, element, timestamp=beam.DoFn.TimestampParam,
+              window=beam.DoFn.WindowParam, * args, **kwargs):
+    if isinstance(element, KeyedWorkItem):
+      # Must be a timer firing.
+      key = element.encoded_key
+    else:
+      key, values = element
+      values = list(values)
+      assert len(values) == 1
+      value = values[0]
+
+    state = self._step_context.get_keyed_state(key)
+    element_state = state.get_state(window, self._element_tag)
+    is_seed_call = not element_state # Initially element_state is an empty list.
```

Comment:
<!--new_thread; commit:7549b5c2ebe2ae47af9066eaf97364a27e828ab5; resolved:1-->
This seems a little fragile to depend on this behavior but I don't see a better alternative right now...  (no action needed)